(latex) separate loading of hyperref from sphinx.sty

jfbu opened the pull request:
new 'hyperref' key in `latex_elements`.

Rationale: user can  add loading of additional packages via `'usepackages'` and `'preamble'` keys to `latex_elements`, but `hyperref`'s loading was hard-coded in sphinx.sty and happens earlier, and this may cause conflicts with attemps by user to employ extra packages. Besides `'hyperref'` key will allow modifying options of `hyperref`.

It is probably better to leave this key un-documented for the time being. I think `'usepackages'` is undocumented currently ?

jfbu committed:
(latex) separate loading of hyperref from sphinx.sty

new 'hyperref' key in ``latex_elements``.

jfbu commented:
I have at long last installed flake8 on my disk so in the future I can do checks before pushing to a PR.

tk0miya commented:
Indeed `hyperref` should be loaded at last. +1

> I think 'usepackages' is undocumented currently ?

The key is reversed key for latex writer. It is filled automatically if extensions calls `Sphinx.add_latex_package()` API.
So users can't set the variable from conf.py.

jfbu commented:
Thanks ! @tk0miya for reviewing and explanations about `'usepackages'` key, I understand now. I replace `\RequirePackage` by `\usepackage` because the code will be placed in produced LaTeX document preamble now. We could consider mentioning`\usepackage{bookmark}` by same author as `hyperref` in documentation of `'hyperref'` key (not yet added to `config.rst`)

jfbu commented:
About location of `'hyperref'` key: we can not put it after `'preamble'` because of `\urlstyle{same}` being part of the default value of the key, and it must not override user set-up.

Thus, I put it before `'preamble'` insertion. In case where user needs to load packages before `hyperref`, the advise is thus that he/she specifies `'hyperref': ''` (empty string) in `latex_elements` and then include contents of default `'hyperref'` key inside the `'preamble'` spec.

Or, have a newer user key `'loadpackages'` which would be located before  `'hyperref'` (itself before `'preamble'`) ?

jfbu committed:
new ``'packages'`` key to ``latex_elements``

serves to load packages after ``sphinx.sty`` itself and before hyperref.

tk0miya commented:
Now we can replace the template of LaTeX writer.
so `loadpackages` key is not needed (at least for the moment).

jfbu commented:
@tk0miya thank you for reviewing! I have reverted the latest commit which inserted a new `'packages'` key to load packages after sphinx but before hyperref. What do you think ?

jfbu committed:
Revert "new ``'packages'`` key to ``latex_elements``"

This reverts commit 9972d729abb076441d9cfe24b637a19e471bab50.

jfbu commented:
rebased and squashed for easying up possibility of merge; leave the two commits about new "'packages'" key and its reverting commit in case this serves in future.

jfbu commented:
thanks @tk0miya for review and advice, I am merging now.
